scripts/cycled-yg-privacy.py

```diff
@@ -0,0 +1,102 @@
+#!/usr/bin/env python3
+from future.utils import iteritems
+
+import random
+
+from jmbase import get_log, jmprint
+from jmbase.support import lookup_appdata_folder
+from jmclient import YieldGeneratorBasic, ygmain, jm_single
+
+# This is a maker for the purposes of generating a yield from held bitcoins
+# while maximising the difficulty of spying on blockchain activity.
+# This is primarily attempted by randomizing all aspects of orders
+# after transactions wherever possible.
+
+# YIELD GENERATOR SETTINGS ARE NOW IN YOUR joinmarket.cfg CONFIG FILE
+# (You can also use command line flags; see --help for this script).
+
+jlog = get_log()
+
+class YieldGeneratorPrivaCycler(YieldGeneratorBasic):
+
+    def __init__(self, wallet_service, offerconfig):
+        super().__init__(wallet_service, offerconfig)
+    
+    def select_input_mixdepth(self, available, offer, amount):
+        """Mixdepths are in cyclic order and we select the mixdepth to
+        maximize the largest interval of non-available mixdepths by chosing
+        the first mixdepth available after the largest such interval.
+        This forces the biggest UTXOs to stay in a bulk of few mixdepths so
+        that the maker can always maximize the size of his orders even when
+        some coins are send from the last to the first mixdepth"""
+        # We sort the available depths for linear scaling of the interval search
+        available = sorted(available.keys())
+        # For an available mixdepth, the smallest interval starting from this mixdepth
+        # containing all the other available mixdepths necessarely ends at the previous
+        # available mixdepth in the cyclic order. The successive difference of sorted
+        # depths is then the length of the largest interval ending at the same mixdepth
+        # without any available mixdepths, modulo the number of mixdepths if 0 is crossed
+        intervals = [(available[i] - available[i-1])%self.wallet_service.mixdepth for i in range(len(available))]
+        # We return the mixdepth value at which the largest interval without
+        # available mixdepths ends. Selecting this mixdepth will send the coinjoin
+        # outputs closer to the others available mixdepths which are after in cyclical order
+        return available[max(range(len(available)), key = intervals.__getitem__)]
+    
+    def create_my_orders(self):
```

Comment:
So, as I understand it, your situation is this: you're inheriting from `YieldGeneratorBasic` and so you must override `create_my_orders` but you don't see a need to change this function from that in `YieldGeneratorPrivacyEnhanced`. I'm guessing you considered inheriting from the latter but discounted it because it isn't exported in `jmclient`? If so we can just export it. Or was there another reason not to do that?
This obviously isn't ideal, as it's a big duplication.

Reply:
> I'm guessing you considered inheriting from the latter but discounted it because it isn't exported in jmclient? If so we can just export it.

Yes, exactly !

I wanted to do use `YieldGeneratorPrivacyEnhanced` class directly but I don't know how exporting it properly. 

Exporting `YieldGeneratorPrivacyEnhanced` in `jmclient` looks fine to me since this is the YG any maker should run for their own privacy but having it along with `YieldGeneratorBasic` also breaks the logic of having a base class of YG in the code that the user overwrites to have the features he wants.

Given this tradeoff, what do you prefer ? I'm fine with both anyway so I follow your opinion